### ISSUE TITLE
loss logged at the end of model.fit() doesn't agree with manually calculated loss

### ISSUE THREAD

**Issue by hadi-ds** [fchollet/keras#3184]
I am trying to build a simple three layer auto-encoder. The .fit() method progressive log shows that by the end of three epochs the training loss drops to 0.0376. I use 'mse' loss. At this point, I expect the network to be able to reconstruct the training data quite well.
However, when I explicitly reconstruct the training data using the trained network, output is way off compared to input. In fact, explicit calculation of cost for training data shows that it is 86.26, compared to 0.0376 reported by .fit() at the end of training (I set aside 10% of data for validation, but it should matter this much). 
So, I wonder what those low loss values logged during training actually correspond to? I supposed those are the total cost function given the state of network (weights) at any moment.

Here is part of my code and its log:

``` python
hidden_dim = 1000
dropout_p = 0.25
l2_lambda = 1e-3
n_epochs = 3
batch_size = 256
AE = autoencoder_nn(n_features, hidden_dim, l2_lambda, dropout_p) 
'''
function definition for 'autoencoder_nn' includes networks design and compilation,
autoencoder.compile(optimizer='adadelta', loss='mse')] 
'''
AE.fit(X_data, X_data,
        nb_epoch=n_epochs, batch_size=batch_size,
        validation_split=0.1, shuffle=True)
```

`
Epoch 1/3 124643/124643 [==============================] - 136s - loss: 1.7028 - val_loss: 0.0032 Epoch 2/3 124643/124643 [==============================] - 128s - loss: 0.2450 - val_loss: 0.0032 Epoch 3/3 124643/124643 [==============================] - 129s - loss: 0.0376 - val_loss: 0.0032
`
And I calculate training error using the following:

``` python
cost = 0.
for row in X_data:
    input = row.reshape((1, X_data.shape[1]))
    pred = AE.predict(input)
    cost += np.linalg.norm(row - pred)**2
print "average cost: ", cost/X_data.shape[0]
```

average cost: 86.2644608377

**Comment by hadi-ds** [comment]
I am trying to build a simple three layer auto-encoder. The .fit() method progressive log shows that by the end of three epochs the training loss drops to 0.0376. I use 'mse' loss. At this point, I expect the network to be able to reconstruct the training data quite well.
However, when I explicitly reconstruct the training data using the trained network, output is way off compared to input. In fact, explicit calculation of cost for training data shows that it is 86.26, compared to 0.0376 reported by .fit() at the end of training (I set aside 10% of data for validation, but it should matter this much). 
So, I wonder what those low loss values logged during training actually correspond to? I supposed those are the total cost function given the state of network (weights) at any moment.

Here is part of my code and its log:    

```python
hidden_dim = 1000
dropout_p = 0.25
l2_lambda = 1e-3
n_epochs = 3
batch_size = 256
AE = autoencoder_nn(n_features, hidden_dim, l2_lambda, dropout_p) 
'''
function definition for 'autoencoder_nn' includes networks design and compilation,
autoencoder.compile(optimizer='adadelta', loss='mse')] 
'''
AE.fit(X_data, X_data,
        nb_epoch=n_epochs, batch_size=batch_size,
        validation_split=0.1, shuffle=True)
```
`Epoch 1/3
124643/124643 [==============================] - 136s - loss: 1.7028 - val_loss: 0.0032
Epoch 2/3
124643/124643 [==============================] - 128s - loss: 0.2450 - val_loss: 0.0032
Epoch 3/3
124643/124643 [==============================] - 129s - loss: 0.0376 - val_loss: 0.0032`  

And I calculate training error using the following:
```python
cost = 0.
for row in X_data:
    input = row.reshape((1, X_data.shape[1]))
    pred = AE.predict(input)
    cost += np.linalg.norm(row - pred)**2
print "average cost: ", cost/X_data.shape[0]
```
average cost: 86.2644608377

**Comment by pianomanu** [comment]
Hi hadi-ds, 

Did you figure out where the problem came from? 
One first guess that you use a ```batch_size``` of 256 in your training, and you predict sample by sample so with ```batch_size=1```
```Try model.predict(X_data,batch_size=256)```

**Comment by hadi-ds** [comment]
Hi there, 

It has been a while since I did this, but as I recall the problem was related to the fact that the output of my network was a vector while `loss='mse'` assumes that output (and the target function network output is compared against) are scalers, i.e., `cost = 1/N \sum_i (y^i_target - y^i_pred)^2`. 
With vectors, it still runs but you can't trust value reported for cost.
I ended up wring my own cost function, `vector_mse=1/N \sum_i ||y^i_target - y^i_pred||^2`, where `y^i_..`'s are vectors here.
